### ISSUE TITLE
[feature] Server-side rendering

### ISSUE THREAD

**Issue by oligot** [lhorie/mithril.js#1356]
tldr; 
It would be great to have a built-in method to ease server-side rendering.

I'm trying to render some components server-side but struggle to find a satisfying solution.

I tried [mithril-node-render](https://github.com/StephanHoyer/mithril-node-render) but it does not play so well with async operations (like fetching some data in the controller of a sub component). There are some propositions regarding async operations [in the linked article](https://gist.github.com/StephanHoyer/bddccd9e159828867d2a) but none that I find satisfying. Moreover, the fact that controller instances are not cached means that the controller is recreated every time we call `render` which is quite different from the way it works in Mithril.

I also found [this other article](https://amir-x.com/how-i-did-mithril-server-rendering-for-some-isomorphic-app-stuff/) but I find it so hacky too.

From what I can understand, the problem is that for the moment, Mithril has a hard dependency on the global `window` object with is ok in the browser but not so ok in Node.js.
I know it's still possible to mock the `window` object but this seems so hacky compared to the simple and elegant API of Mithril.

What would be really great for server-side rendering is a method that is able to render a component as a string (or a stream), as can be found in [React](https://facebook.github.io/react/docs/top-level-api.html#reactdomserver.rendertostring) or [Vue.js](https://vuejs.org/guide/ssr.html). Do you think it's possible to add such a method to Mithril and if not, what are your recommendations to render components server-side ?

**Comment by isiahmeadows** [comment]
The rewrite also has a somewhat hard dependency on the `window`, but it's probably easier to stub out.

Oh, and [you're not the first to think of it](https://github.com/lhorie/mithril.js/issues/856). It's just that removing the hard dependency is far from trivial, and generally requires a rewrite to avoid.

**Comment by isiahmeadows** [comment]
@StephanHoyer is [currently working on the server side](https://github.com/StephanHoyer/mithril-node-render/tree/rewrite), and the now-closed #977 was a broken attempt to make the revival work on the browser end ([it needs a bit of design work first](https://github.com/lhorie/mithril.js/pull/1308#issuecomment-262445253)).

**Comment by oligot** [comment]
Ok, thanks for the info.

**Comment by balupton** [comment]
For reference, the earlier link to mithril-node-render doesn't work, as the branch is deleted, so https://github.com/StephanHoyer/mithril-node-render is where to go now.

For the topic - I'd love to see server-side-rendering support built into Mithril. It seems the last piece of the puzzle missing. Looking at the source of https://github.com/StephanHoyer/mithril-node-render it doesn't seem the changes are that large. Is there any plans or roadmap for built-in support for server-side-rendering now that Mithril v1.0.0 is out the door?

**Comment by StephanHoyer** [comment]
Mithril-node-render should work just fine with 1.0


Am 07.02.2017 21:20 schrieb "Benjamin Lupton" <notifications@github.com>:

**Comment by tivac** [comment]
If `mithril-node-render` isn't the officially-recommended solution in the docs it probably should be.

I don't have any desire to bundle it alongside the main library though, I think a separate package is the perfect place for functionality like that.

**Comment by balupton** [comment]
well, the question is why is the core designed in such a way that it is dependent on an actual DOM - it seems that `.render` should be fine without a DOM, and that only `.mount` should need a DOM

**Comment by isiahmeadows** [comment]
@balupton What would `.render` conceptually do without a DOM? If you mean for it to render to a string, it doesn't make sense to return two different things based on the presence of a native DOM (it could be injected later with JSDOM) or not.

**Comment by balupton** [comment]
@isiahmeadows that's a good point - took a look at the source, you are definitely right

so is server-side-rendering an officially supported use case (even if accomplished via a ecosystem package)? as it seems a common one - would be nice if there was a website page mentioning the stance on it

**Comment by tivac** [comment]
Don't want to speak for @lhorie but we've been rendering mithril on the server using `mithril-node-render` for almost two years at this point. It's an excellent piece of software.

**Comment by StephanHoyer** [comment]
It also supports async rendering now

**Comment by dineshselvantdm** [comment]
Is there any checksum mechanism for serverside rendered HTML using mithril-node-render and mithril to avoid re-rendering in client? @lhorie @StephanHoyer

**Comment by isiahmeadows** [comment]
@dineshselvantdm No, and it's unlikely that would provide much benefit (React even stopped using it), but there has been past discussion on optimizing mounting to a pre-rendered HTML tree. Could you file a new issue for that?

**Comment by dineshselvantdm** [comment]
@isiahmeadows Thanks! Filed the issue #1838.

**Comment by isiahmeadows** [comment]
> React v15 still uses data-react-id for checksum.

That's probably for verification and compatibility with previous versions, not for actual diffing and rebuilding.

**Comment by dineshselvantdm** [comment]
**data-react-checksum** is still used by react, which uses Alder-32 checksum algorithm for comparing the serverside HTML with first VDOM in client.
